File: Stripe/UINavigationController+Stripe_Completion.m
//
//  UINavigationController+Stripe_Completion.m
//  Stripe
//
//  Created by Jack Flintermann on 3/23/16.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

#import "UINavigationController+Stripe_Completion.h"

@implementation UINavigationController (Stripe_Completion)

- (void)stp_pushViewController:(UIViewController *)viewController
                      animated:(BOOL)animated
                    completion:(STPVoidBlock)completion {
    [CATransaction begin];
    [CATransaction setCompletionBlock:completion];
    [self pushViewController:viewController animated:animated];
    [CATransaction commit];
}

- (void)stp_popViewControllerAnimated:(BOOL)animated
                           completion:(STPVoidBlock)completion {
    [CATransaction begin];
    [CATransaction setCompletionBlock:completion];
    [self popViewControllerAnimated:animated];
    [CATransaction commit];
}

- (void)stp_popToViewController:(UIViewController *)viewController
                       animated:(BOOL)animated
                     completion:(STPVoidBlock)completion {
    [CATransaction begin];
    if (completion) {
        [CATransaction setCompletionBlock:completion];
    }
    [self popToViewController:viewController animated:animated];
    [CATransaction commit];
}

@end

File: Stripe/UINavigationController+Stripe_Completion.swift
//
//  UINavigationController+Stripe_Completion.swift
//  Stripe
//
//  Created by Jack Flintermann on 3/23/16.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

import UIKit

// See http://stackoverflow.com/questions/9906966/completion-handler-for-uinavigationcontroller-pushviewcontrolleranimated/33767837#33767837 for some discussion around why using CATransaction is unreliable here.

extension UINavigationController {
  func stp_push(
    _ viewController: UIViewController?,
    animated: Bool,
    completion: @escaping STPVoidBlock
  ) {
    if let viewController = viewController {
      pushViewController(viewController, animated: animated)
    }
    if transitionCoordinator != nil && animated {
      transitionCoordinator?.animate(alongsideTransition: nil) { _ in
        completion()
      }
    } else {
      completion()
    }
  }

  func stp_popViewController(
    animated: Bool,
    completion: @escaping STPVoidBlock
  ) {
    popViewController(animated: animated)
    if transitionCoordinator != nil && animated {
      transitionCoordinator?.animate(alongsideTransition: nil) { _ in
        completion()
      }
    } else {
      completion()
    }
  }

  @objc(stp_popToViewController:animated:completion:) func stp_pop(
    to viewController: UIViewController?,
    animated: Bool,
    completion: @escaping STPVoidBlock
  ) {
    if let viewController = viewController {
      popToViewController(viewController, animated: animated)
    }
    if transitionCoordinator != nil && animated {
      transitionCoordinator?.animate(alongsideTransition: nil) { _ in
        completion()
      }
    } else {
      completion()
    }
  }
}
